### PR TITLE
tests: Add initial swarm tests

### LINKED ISSUE #198: tests: Add initial swarm tests
Submit initial swarm tests in order to address https://github.com/kata-containers/runtime/issues/174

### PR DESCRIPTION
**Pull request by GabyCT**
This will enable basic swarm tests.

Fixes #198

Signed-off-by: Gabriela Cervantes <gabriela.cervantes.tellez@intel.com>

### CODE REVIEW

File: integration/swarm/swarm.bats [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		--name "${SERVICE_NAME}" --replicas $number_of_replicas \
+		--publish 8080:80 "${nginx_image}" sh -c "$nginx_command"
+	running_regex='Running\s+\d+\s(seconds|minutes)\s+ago'
+	for i in $(seq "$timeout") ; do
+		docker service ls --filter name="$SERVICE_NAME"
+		replicas_running=$(docker service ps "$SERVICE_NAME" | grep -P "${running_regex}"  | wc -l)
+		if [ "$replicas_running" -ge "$number_of_replicas" ]; then
+			break;
+		fi
+		sleep
```
Comment (chavafg) [inline integration/swarm/swarm.bats]: missing value on sleep?

Comment (devimc) [inline integration/swarm/swarm.bats]: missing operand

File: integration/swarm/swarm.bats [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	# here we are checking that each replica has two interfaces
+	# and they should be always eth0 and eth1
+	REPLICA_ID=$(docker ps -q)
+	docker exec ${REPLICA_ID} sh -c "ip route show | grep -E eth0 && ip route show | grep -E eth1" > /dev/null
+}
+
+teardown() {
+	docker service remove "${SERVICE_NAME}"
+	docker swarm leave --force
+	check_processes ${SHIM_PATH}
```
Comment (chavafg) [inline integration/swarm/swarm.bats]: should we also check for qemu and proxy processes?

File: integration/swarm/swarm.bats [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+			break;
+		fi
+	done
+	docker swarm init ${swarm_interface_arg}
+	nginx_command="hostname > /usr/share/nginx/html/hostname; nginx -g \"daemon off;\""
+	docker service create \
+		--name "${SERVICE_NAME}" --replicas $number_of_replicas \
+		--publish 8080:80 "${nginx_image}" sh -c "$nginx_command"
+	running_regex='Running\s+\d+\s(seconds|minutes)\s+ago'
+	for i in $(seq "$timeout") ; do
```
Comment (chavafg) [inline integration/swarm/swarm.bats]: you should check the `$i` value inside the loop to see if we have reached the timeout and if so, return 1?

File: integration/swarm/swarm.bats [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+		docker service ls --filter name="$SERVICE_NAME"
+		replicas_running=$(docker service ps "$SERVICE_NAME" | grep -P "${running_regex}"  | wc -l)
```
Comment (chavafg) [inline integration/swarm/swarm.bats]: you can use `grep -c -P "${running_regex}" ` and remove `wc -l`

**Comment by GabyCT** [comment]
@chavafg the ci is failing with the following error .ci/teardown.sh "$WORKSPACE"
[kata-containers-tests-ubuntu-16-04-PR] $ /bin/bash /tmp/jenkins2389151152929419399.sh
sudo: kata-collect-data.sh: command not found
~/jenkins_slave/workspace/kata-containers-tests-ubuntu-16-04-PR ~/jenkins_slave/workspace/kata-containers-tests-ubuntu-16-04-PR/go/src/github.com/kata-containers/tests
gzip: kata-collect-data_*: No such file or directory

Any idea?

**Comment by GabyCT** [comment]
@jodh-intel thanks :)

**Comment by chavafg** [comment]
I think that is not the root cause, above you can see:

```
~/jenkins_slave/workspace/kata-containers-tests-ubuntu-16-04-PR/go/src/github.com/kata-containers/osbuilder/image-builder ~/jenkins_slave/workspace/kata-containers-tests-ubuntu-16-04-PR/go/src/github.com/kata-containers/tests
ERROR: ../rootfs-builder/rootfs is not a directory
Build step 'Execute shell' marked build as failure
```

**Comment by chavafg** [comment]
maybe because of: https://github.com/kata-containers/osbuilder/pull/63 ?

**Comment by jodh-intel** [comment]
Yes - that does look like an issue. @erick0z - could you take a look?

**Comment by chavafg** [comment]
testing locally, the old `rootfs` directory is now called `rootfs-Clear`, because we are using clearlinux as the `OS_BUILDER_DISTRO`:
```
OSBUILDER_DISTRO=${OSBUILDER_DISTRO:-clearlinux}  
```

```
~/go/src/github.com/kata-containers/osbuilder/image-builder ~/go/src/github.com/kata-containers/tests
+ sudo -E AGENT_INIT=no USE_DOCKER=true ./image_builder.sh ../rootfs-builder/rootfs
ERROR: ../rootfs-builder/rootfs is not a directory
```

So I can change the name of the directory, but as we will test more distros in the future I am not sure what will be the best option here.
Maybe the directory should be called `rootfs-clearlinux` instead of `rootfs-Clear`? 

cc @erick0z, @jcvenegas @jodh-intel @devimc

**Comment by chavafg** [comment]
@jodh-intel, seems like it will need to be an absolute path, and after I changed to
```
export ROOTFS_DIR="${GOPATH}/src/${osbuilder_repo}/rootfs-builder/rootfs"
```

I am still getting the error. Seems like the `rootfs` directory is not being populated.

**Comment by erick0z** [comment]
@chavafg The latest changes introduced to the osbuilder repo changes the resulting rootfs directory name, so the `../rootfs-builder/rootfs` directory won't exist anymore. It should exist, for instance: `../rootfs-builder/rootfs-Clear`.

Have you tried the following?
`sudo -E AGENT_INIT=no USE_DOCKER=true ./image_builder.sh ../rootfs-builder/rootfs-Clear`

Other option is, when building the rootfs:
` sudo -E PATH=$PATH DISTRO=clearlinux DISTRO_ROOTFS="${GOPATH}/src/${osbuilder_repo}/rootfs-builder/rootfs" make rootfs`

**Comment by chavafg** [comment]
@GabyCT opened https://github.com/kata-containers/tests/pull/212, which solve the issue, after it is merged, please rebase your branch

**Comment by GabyCT** [comment]
@jodh-intel  thanks for the help and @chavafg thanks for solving the issue with the CI, the problem was solved

**Comment by GabyCT** [comment]
@chavafg changes applied

File: integration/swarm/swarm.bats [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+		--name "${SERVICE_NAME}" --replicas $number_of_replicas \
+		--publish 8080:80 "${nginx_image}" sh -c "$nginx_command"
+	running_regex='Running\s+\d+\s(seconds|minutes)\s+ago'
+	for i in $(seq "$timeout") ; do
+		docker service ls --filter name="$SERVICE_NAME"
+		replicas_running=$(docker service ps "$SERVICE_NAME" | grep -c -P "${running_regex}")
+		if [ "$replicas_running" -ge "$number_of_replicas" ]; then
+			break
+		fi
+		sleep 1
```
Comment (chavafg) [inline integration/swarm/swarm.bats]: can you add below something like:
[ "$i" == "$timeout" ] && return 1
which would mean that we are reaching the timeout and no replicas were running, then we need to fail.

**Pull request merged**